src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,374 @@
+
+void segment_upload::throw_if_not_initialized(std::string_view caller) const {
+    if (_stream.has_value() == false) {
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format,
+          "Segment upload is not initialized {} to {}",
+          _ntp,
+          caller));
+    }
+}
+
+ss::future<result<std::unique_ptr<segment_upload>>>
+segment_upload::make_segment_upload(
+  ss::lw_shared_ptr<cluster::partition> part,
+  inclusive_offset_range range,
+  size_t read_buffer_size,
+  ss::scheduling_group sg,
+  model::timeout_clock::time_point deadline) {
+    std::unique_ptr<segment_upload> upl(
+      new segment_upload(part, read_buffer_size, sg));
+
+    auto res = co_await upl->initialize(range, deadline);
+    if (res.has_failure()) {
+        co_return res.as_failure();
+    }
+    co_return std::move(upl);
+}
+
+ss::future<result<std::unique_ptr<segment_upload>>>
+segment_upload::make_segment_upload(
+  ss::lw_shared_ptr<cluster::partition> part,
+  size_limited_offset_range range,
+  size_t read_buffer_size,
+  ss::scheduling_group sg,
+  model::timeout_clock::time_point deadline) {
+    std::unique_ptr<segment_upload> upl(
+      new segment_upload(part, read_buffer_size, sg));
+
+    auto res = co_await upl->initialize(range, deadline);
+    if (res.has_failure()) {
+        co_return res.as_failure();
+    }
+    co_return std::move(upl);
+}
+
+ss::future<result<void>> segment_upload::initialize(
+  inclusive_offset_range range, model::timeout_clock::time_point deadline) {
+    auto holder = _gate.hold();
+    auto params = co_await compute_upload_parameters(range, deadline);
+    if (params.has_failure()) {
+        co_return params.as_failure();
+    }
+    _params = std::move(params.value());
+    // Create a log reader config to scan the uploaded offset
+    // range. We should skip the batch cache.
+    storage::log_reader_config reader_cfg(
+      range.base, range.last, priority_manager::local().archival_priority());
+    reader_cfg.skip_batch_cache = true;
+    reader_cfg.skip_readers_cache = true;
+    vlog(_ctxlog.debug, "Creating log reader, config: {}", reader_cfg);
+    auto reader = co_await _part->make_reader(reader_cfg);
+    _stream = make_reader_input_stream(
+      _ntp,
+      std::move(reader),
+      _rd_buffer_size,
+      range,
+      model::time_until(deadline));
+    co_return outcome::success();
+}
+
+ss::future<result<void>> segment_upload::initialize(
+  size_limited_offset_range range, model::timeout_clock::time_point deadline) {
+    auto holder = _gate.hold();
+    auto params = co_await compute_upload_parameters(range, deadline);
+    if (params.has_failure()) {
+        co_return params.as_failure();
+    }
+    _params = std::move(params.value());
+    // Create a log reader config to scan the uploaded offset
+    // range. We should skip the batch cache.
+    storage::log_reader_config reader_cfg(
+      params.value().offsets.base,
+      params.value().offsets.last,
```

Comment:
`params` is moved before this so we should use `_params`

Also I see we use the fields from params here, so maybe it makes sense to use the computed range in case of `size_limited_offset_range` and in the other case it is fine to use the input range directly (in context of my question [here](https://github.com/redpanda-data/redpanda/pull/16999/commits/db59a2db74a5147b73fd7e9e326f6621a6bb623c#r1524389189)).